[WIP] Initial support for using AWS config files for storing credentials.

davidewatson opened the pull request:
Using this will allow us to include terraform.tfvars infrastructure
directly in the test logs. Note that the old method of specifying
credentials in tfvar files is still there for backwards compatibility.

Before this is merged, the documentation should to be updated (though the change is backwards compatible), and I need to articulate why I think credential files are a more flexible solution than environment variables, especially in view of a Jenkins server which may need to support more than one AWS account.

davidewatson commented on bin/kraken-up.sh:
```diff
@@ -26,9 +26,15 @@ if docker inspect ${KRAKEN_CONTAINER_NAME} &> /dev/null; then
+
+# This will copy the credentials to the home directory of the default user,
+# which is ubuntu.
+run_command "docker-machine scp -r ~/.aws ${KRAKEN_DOCKER_MACHINE_NAME}":
```
This line may need to be made conditional on whether the user is using docker-machine...

davidewatson commented:
We might even be able to support environment variables and configuration files by doing something like this:

```-run_command "docker run -d --name ${KRAKEN_CONTAINER_NAME} --volumes-from kraken_data ${KRAKEN_CONTAINER_IMAGE_NAME}
-  bash -c \"/opt/kraken/terraform-down.sh --clustertype ${KRAKEN_CLUSTER_TYPE} --clustername ${KRAKEN_CLUSTER_NAME} --terraform-retries ${TERRAFORM_RETRIES}\""
+run_command "docker run -d --name ${KRAKEN_CONTAINER_NAME} \
+        -e AWS_ACCESS_KEY_ID=${AWS_ACCESS_KEY} \
+        -e AWS_SECRET_ACCESS_KEY=${AWS_SECRET_ACCESS_KEY} \
+        -v /home/ubuntu/.aws:/root/.aws/ \
+        -v /var/run:/ansible \
+        --volumes-from kraken_data ${KRAKEN_CONTAINER_IMAGE_NAME} \
+  bash -c \"/opt/kraken/terraform-down.sh \
+        --clustertype ${KRAKEN_CLUSTER_TYPE} \
+        --clustername ${KRAKEN_CLUSTER_NAME} \
+        --terraform-retries ${TERRAFORM_RETRIES}\""```

davidewatson commented:
please retest this

davidewatson commented:
retest this please

spiffxp commented on terraform/aws/variables.tf:
```diff
@@ -5,10 +5,22 @@ variable "cluster_name" {
+
+variable "aws_shared_credentials_file" {
+  description = "This is the path to the shared credentials file. If this is not set and a profile is specified, ~/.aws/credentials will be used."
+  default = ""
```
why not just set the default here?

davidewatson replied:
I agreed that explicit is better than implicit, but in this case I couldn't find a way to get terraform to expand tildes, and since I can't know the absolute path before hand...

spiffxp commented on terraform/aws/aws.tf:
```diff
@@ -1,6 +1,8 @@
   secret_key  = "${var.aws_secret_key}"
+  shared_credentials_file = "${var.aws_shared_credentials_file}"
+  profile     = "${var.aws_profile}"
   region      = "${var.aws_region}"
```
I would expect this profile block to no longer have credentials stored inside of it if we're using an external credentials file?  otherwise I believe these values get saved in .tfstate and we're thus leaking secrets

davidewatson replied:
All that is saved is the name of the file and profile, not the actual credentials.

spiffxp commented on bin/kraken-down.sh:
```diff
@@ -25,8 +25,22 @@ if ! docker inspect kraken_data &> /dev/null; then
+  AWS_CONFIG_DIRECTORY="/home/ubuntu/.aws"
+else
+  AWS_CONFIG_DIRECTORY="${HOME}/.aws"
+fi
```
pull this block into utils.sh and you avoid copy-pasta between -up.sh and -down.sh

davidewatson replied:
Agreed.

spiffxp commented on bin/build/Dockerfile:
```diff
@@ -1,13 +1,14 @@
-ENV ANSIBLE_VERSION 2.0.0.2
-ENV AWSCLI_VERSION 1.10.4
+# Environment variables. Note that a single ENV line is preferred because it
+# produces a single cache layer.
```
this is the sort of info I'd expect to see in the commit message FWIW, not sure it necessarily needs to be a comment in the Dockerfile (we don't comment on this trick everywhere we use large `RUN` statements for example)

davidewatson replied:
Will fix.

spiffxp commented on README.md:
```diff
@@ -16,22 +16,35 @@ For example, for aws cluster named "my_neato_cluster", the file will be:
+
+to specify which profile to use. By default the profile name will be "default".
+
+2) Provide your credentials directly in the terraform.tfvars files
```
I was under the impression we were going to drop support for this approach

spiffxp commented:
These changes are only in .sh variants, do we bother keeping the .ps1 variants in sync?  Do we note they're falling behind?  Do we outright remove them?

spiffxp commented on README.md:
```diff
@@ -45,23 +58,22 @@ For better performance, you should consider adding and modifing the following co
 ### Cluster Services
 
-Kraken supports turnkey deployment of a number of useful cluster services, via the [kraken-services](https://github.com/samsung-ag/kraken-services) repository.  Don't see a service you want in our repo?  You can use your own!
+Kraken supports turnkey deployment of a number of useful cluster services, via the [kraken-services](https://github.com/samsung-cnct/kraken-services) repository.  Don't see a service you want in our repo?  You can use your own!
```
good catch

spiffxp replied:
testing slack notifications

spiffxp commented on bin/kraken-up.sh:
```diff
@@ -26,10 +26,26 @@ if docker inspect ${KRAKEN_CONTAINER_NAME} &> /dev/null; then
 fi
 
-# run cluster up
+# Create container which will manage infrastructure.
```
'infrastructure' is way too generic a term here, maybe 'a single kubernetes cluster'?

maratoid commented on terraform/aws/aws.tf:
```diff
@@ -1,6 +1,8 @@
 provider "aws" {
   access_key  = "${var.aws_access_key}"
```
No need for aws_access_key in here or in the vars file any longer

maratoid commented on terraform/aws/aws.tf:
```diff
@@ -1,6 +1,8 @@
 provider "aws" {
   access_key  = "${var.aws_access_key}"
   secret_key  = "${var.aws_secret_key}"
```
No need for aws_secret_key in here or in the vars file any longer

maratoid commented on bin/kraken-up.sh:
```diff
@@ -26,10 +26,26 @@ if docker inspect ${KRAKEN_CONTAINER_NAME} &> /dev/null; then
+
+# Determine location of AWS configuration directory
+if [ "${KRAKEN_NATIVE_DOCKER}" = false ] ; then
+  run_command "docker-machine scp -r ~/.aws ${KRAKEN_DOCKER_MACHINE_NAME}": || true
```
We probably should not assume that the credentials file we want to use for docker-machine is in ~/.aws

davidewatson replied:
Good call. Will add an optional parameter to kraken-up.sh.

davidewatson replied:
I've implemented a --provider-credential-directory option to kraken-up.sh.

Note though, AWS tools and docker always expect it to be located at ~/.aws/, only terraform allows it to vary. This is not really a problem for the docker-machine case, we can just always put the credentials in ~/.aws. But for native docker this is a problem, because it could conflict with other credentials the user stores there. My proposal will be to disallow --provider-credential-directory in the native docker case.

maratoid commented on bin/kraken-up.sh:
```diff
@@ -26,10 +26,26 @@ if docker inspect ${KRAKEN_CONTAINER_NAME} &> /dev/null; then
+# Determine location of AWS configuration directory
+if [ "${KRAKEN_NATIVE_DOCKER}" = false ] ; then
+  run_command "docker-machine scp -r ~/.aws ${KRAKEN_DOCKER_MACHINE_NAME}": || true
+  AWS_CONFIG_DIRECTORY="/home/ubuntu/.aws"
```
We probably should not assume that our docker-machine driver is an ubuntu VM with 'ubuntu' user

maratoid commented on bin/kraken-up.sh:
```diff
@@ -26,10 +26,26 @@ if docker inspect ${KRAKEN_CONTAINER_NAME} &> /dev/null; then
+  run_command "docker-machine scp -r ~/.aws ${KRAKEN_DOCKER_MACHINE_NAME}": || true
+  AWS_CONFIG_DIRECTORY="/home/ubuntu/.aws"
+else
+  AWS_CONFIG_DIRECTORY="${HOME}/.aws"
```
We probably should not assume that the credentials file we want to use is in ~/.aws

maratoid commented:
As far as ps1 changes go - I hate to say this, but maybe we should talk this over? We've been keeping functionalities in sync up until now - but maybe we should deprecate them? Especially now that Win 10 has bash

davidewatson commented:
I believe I have addressed all comments except for the decision of whether to drop Powershell support. I am closing this PR and will open a new one with better hygiene.
